src/foam/mlang/mlang.js

```diff
@@ -4715,24 +4716,39 @@ foam.CLASS({
       name: 'partialEval',
       type: 'foam.mlang.Expr',
       javaCode: `
-        List<Double> list = new ArrayList<>();
+        if ( getArgs().length == 0 ) return this;
+        if ( getArgs().length == 1 ) return getArgs()[0].partialEval();
+
+        List<Double> valList = new ArrayList<>();
+        List<Expr>   argList = new ArrayList<>();
         for ( var arg : getArgs() ) {
           arg = arg.partialEval();
           if ( arg instanceof Constant ) {
             var value = ((Number) arg.f(this)).doubleValue();
-            list.add(value);
+            valList.add(value);
+          } else {
+            argList.add(arg);
           }
         }
 
-        if ( list.size() == getArgs().length ) {
-          var result = list.stream().reduce(this::reduce).get();
+        var result = valList.stream().reduce(this::reduce);
+        if ( result.isPresent() ) {
+          var value = result.get();
+          if ( ! Double.isFinite(value) ) return new Constant(value);
```

Comment:
Do you want the '!' here?

Reply:
Yes, it's to early return if the partial eval could return Infinity and NaN. Arithmetic on Infinity or NaN is fruitless.